### ISSUE TITLE
Advice on `num-documents-parallel-load` configuration

### ISSUE THREAD

**Issue by Bug-Reaper** [typesense/typesense#1133]
## Advice on `num-documents-parallel-load` configuration

We run some typesense databases with a single collection of ~6.5 million documents and the startup is quite slow (around 2hrs or so)

Wondered if someone could advise how to configure (`num-documents-parallel-load`). CPU usage seems quite low during start-up. Is it safe to maybe bump this from 1000 => 10,000 on a 24 core / 32 core machine w/128Gib of ram?

<img width="740" alt="Screenshot 2023-08-02 at 10 41 41 PM" src="https://github.com/typesense/typesense/assets/31227781/d45d26cc-337a-4051-87de-24e5a636a215">

## Metadata

**Typesense Version**:
0.24.1 - AMD-64 Binary

**OS**:
Linux (Arch Linux)

**Comment by kishorenc** [comment]
Yes you can bump up this number safely for your given configuration. 

Are you using vector index or have very large text? Generally 6M docs should not take so long.

**Comment by Bug-Reaper** [comment]
As always appreciate the response Kishorenc. We do indeed have a vector_index property of length 256. And we may add other vector_index props in the future.

Is there any danger in cranking this number as high as possible until failure or do you know if it would fail gracefully (no data corruption) should the number be too high for a given system config?

**Comment by kishorenc** [comment]
Higher `num-documents-parallel-load` means higher number of documents buffered in memory during the indexing phase at startup (doesn't affect live indexing).  So if you have enough memory you could increase it further but I suspect that you are already at a point of diminishing returns now.

The other parameter to look at is `thread-pool-size` which controls the total number of threads available. The default value for this is `proc_count * 8`. This should already be sufficiently high given 24 cores, but I just wanted to mention this anyway.

**Comment by Bug-Reaper** [comment]
> Higher `num-documents-parallel-load` means higher number of documents buffered in memory during the indexing phase at startup (doesn't affect live indexing). So if you have enough memory you could increase it further but I suspect that you are already at a point of diminishing returns now.
> 
> The other parameter to look at is `thread-pool-size` which controls the total number of threads available. The default value for this is `proc_count * 8`. This should already be sufficiently high given 24 cores, but I just wanted to mention this anyway.

So technically current set-up is a 12 core machine w/24 threads. Want to confirm for the formula `proc_count * 8` on thread pool size, should we plugin`24 * 8` or `12 * 8`?

Even on previous tests w/ 72 thread (32 core) machine and very high thread-pool-size the initialization was pretty slow (2hrs). Majority of threads sat at around 1-0% usage for the init time too. With our big memory overhead I may try to increase `num-documents-parallel-load` on our next boot.

**Comment by Bug-Reaper** [comment]
Still curious:

1. Does `proc_count` refer the number of cores or the number of threads?
2. If we crank `num-documents-parallel-load` too high and get an OOM memory during init, is there risk of our database becoming corrupted?

**Comment by kishorenc** [comment]
1. `proc_count` refers to number of hyper threads. 
2. No, database is always restored from a snapshot and init does not write to disk, only to memory, so this will not happen.
